File: client/src/androidTest/java/com/jraska/github/client/http/MockingResponsesHttpComponent.java
package com.jraska.github.client.http;

import android.support.test.espresso.IdlingRegistry;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public final class MockingResponsesHttpComponent implements HttpComponent {
  private static final String NETWORK_ERROR_MESSAGE = "You are trying to do network requests in tests you naughty developer!";

  private final Retrofit retrofit;

  private MockingResponsesHttpComponent(Retrofit retrofit) {
    this.retrofit = retrofit;
  }

  @Override public Retrofit retrofit() {
    return retrofit;
  }

  public static MockingResponsesHttpComponent create() {
    RxHttpIdlingResourceFactory idlingResourceFactory = RxHttpIdlingResourceFactory.create();
    IdlingRegistry.getInstance().register(idlingResourceFactory.idlingResource());

    Retrofit retrofit = new Retrofit.Builder()
      .baseUrl("https://api.github.com")
      .addCallAdapterFactory(idlingResourceFactory)
      .addConverterFactory(GsonConverterFactory.create())
      .client(failingClient())
      .build();

    return new MockingResponsesHttpComponent(retrofit);
  }

  private static OkHttpClient failingClient() {
    return new OkHttpClient.Builder()
      .addInterceptor(new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BASIC))
      .addInterceptor(new MockResponsesInterceptor())
      .addInterceptor(chain -> {
        throw new UnsupportedOperationException(NETWORK_ERROR_MESSAGE);
      })
      .build();
  }
}

File: client/src/androidTest/java/com/jraska/github/client/http/MockingResponsesHttpComponent.kt
package com.jraska.github.client.http

import android.support.test.espresso.IdlingRegistry
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class MockingResponsesHttpComponent private constructor(private val retrofit: Retrofit) : HttpComponent {

  override fun retrofit(): Retrofit {
    return retrofit
  }

  companion object {
    private val NETWORK_ERROR_MESSAGE = "You are trying to do network requests in tests you naughty developer!"

    fun create(): MockingResponsesHttpComponent {
      val idlingResourceFactory = RxHttpIdlingResourceFactory.create()
      IdlingRegistry.getInstance().register(idlingResourceFactory.idlingResource())

      val retrofit = Retrofit.Builder()
        .baseUrl("https://api.github.com")
        .addCallAdapterFactory(idlingResourceFactory)
        .addConverterFactory(GsonConverterFactory.create())
        .client(failingClient())
        .build()

      return MockingResponsesHttpComponent(retrofit)
    }

    private fun failingClient(): OkHttpClient {
      return OkHttpClient.Builder()
        .addInterceptor(HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BASIC))
        .addInterceptor(MockResponsesInterceptor())
        .addInterceptor { _ -> throw UnsupportedOperationException(NETWORK_ERROR_MESSAGE) }
        .build()
    }
  }
}
